docs/source/distributed.pipelining.rst

```diff
@@ -4,139 +4,216 @@
+and their data-flow relationship.
+The runtime executes the pipeline stages on different devices in parallel,
+handling things like micro-batch splitting, scheduling, communication, and
+gradient propagation, etc.
 
-While promising for scaling, pipelining is often difficult to implement, requiring intrusive code changes to model code and difficult-to-implement runtime orchestration code. ``torch.distributed.pipelining`` aims to provide **a toolkit that does said things automatically to allow high-productivity scaling of models.** It consists of a **compiler** and a **runtime** stack for easy pipelining of PyTorch models. In particular, it provides the following features:
+Overall, the ``pipelining`` package provides the following features:
 
-* Splitting of model code based on your specification. The goal is for the user to provide model code as-is to the system for parallelization, without having to make heavyweight modifications to make parallelism work. The specification is also simple.
-* Support for rich pipeline scheduling paradigms, including GPipe, 1F1B, Interleaved 1F1B and Looped BFS. It will be also easy to customize your own schedule under this framework.
-* First-class support for cross-host pipeline parallelism, as this is where PP is typically used (over slower interconnects).
-* Composability with other PyTorch parallel schemes such as data parallelism (DDP, FSDP) or tensor  parallelism (overall, known as "3d parallelism").
+* Splitting of model code based on your specification. The goal is for the user
+  to provide model code as-is, without having to make heavyweight modifications
+  to make parallelism work. The specification is simple.
+* Support for rich pipeline scheduling paradigms, including GPipe, 1F1B,
+  Interleaved 1F1B and Looped BFS. It will be also easy to customize your own
+  schedule with this framework too.
+* First-class support for cross-host pipeline parallelism, as this is where PP
+  is typically used (over slower interconnects).
+* Composability with other PyTorch parallel schemes such as data parallelism
+  (DDP, FSDP) or tensor  parallelism (overall, known as "3d parallelism").
 
-Examples
-********
 
-In the `PiPPy <https://github.com/pytorch/PiPPy>`_ repo where this package is migrated from, we provide rich examples based on realistic models. In particular, we show how to apply pipelining without any model code change. You can refer to the `HuggingFace examples directory <https://github.com/pytorch/PiPPy/tree/main/examples/huggingface>`_. Popular examples include: `GPT2 <https://github.com/pytorch/PiPPy/tree/main/examples/huggingface/pippy_gpt2.py>`_, and `LLaMA <https://github.com/pytorch/PiPPy/tree/main/examples/llama>`_.
+Step 1: Choosing the Frontend that Fits Your Need
+*************************************************
 
-Techniques Explained
-********************
+The ``pipelining`` package provides two frontends for two different use cases.
```

Comment:
I think there is still a blurry line between the two frontends, where you may have a full model but still choose to use the manual frontend due to tracer limitations.  How do we approach this coherently?